### PR TITLE
[WIP] Pack and unpack of table_view for serialization

### LINKED ISSUE #3793: [FEA] Function to "pack" a table into a single buffer
**Is your feature request related to a problem? Please describe.**

I would like to be able to create a copy of a `table_view` where the storage for the constituent column is "packed" into a single allocation. 

**Describe the solution you'd like**

An API something like this:
```
std::pair<rmm::device_buffer, table_view> pack(table_view const& t);
```

Where the returned `device_buffer` contains a copy of the contents of `t` in a single allocation and the returned `table_view` points into that buffer. 

**Additional context**

This functionality is effectively already developed for `contiguous_split` in the[ `alloc_and_copy` ](https://github.com/rapidsai/cudf/blob/930810ae3f26fe49eddee70f9aedc6ff10f3ba43/cpp/src/copying/contiguous_split.cu#L175) function. The ask is just to wrap and expose this as a public API.

### PR DESCRIPTION
**Pull request by devavret**
Closes #3793

### CODE REVIEW

File: cpp/src/copying/pack.cpp [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    serial_column._type,
+    serial_column._size,
+    data_ptr,
+    null_mask_ptr,
+    UNKNOWN_NULL_COUNT,
+    0,
+    children);
+}
+
+std::vector<column_view> get_columns(cudf::size_type num_columns,
```
Comment (jrhemstad) [inline cpp/src/copying/pack.cpp:78]: Similar to serialization,  I suggest renaming this to `deserialize_columns` (plural) and  `deserialize_column` (singular) to just `deserialize`. 

Then, lets try to make it iterator based to avoid needing to carry around the `current_index`. This one was a little trickier than `serialize`. 

I _think_ something like this should work:
```
template <typename Input>
pair<vector<column_view>, Input> deserialize_columns(Input serialized_begin, Input serialized_end, unsigned char const * base_ptr){
  vector<column_view> deserialized;
  while(serialized_begin != serialized_end){
      auto current = *(serialized_begin++);
      // deserialize any children of the current column
      auto [children, serialized_begin] = deserialize_columns(serialized_begin,  serialized_begin + current.num_children, base_ptr); // advances serialized_begin by the number of descendants 
     deserialized.push_back(deserialize(current, children, base_ptr)); 
   }
   return {deserialized, serialized_begin}; 
}
```

Maybe @codereport or @harrism can come up with a better algorithm here.

In short, the problem is knowing how far to advance the `serialized_begin` iterator each iteration of the `while` loop. 

The lay out is `[col0, col0.child0, col0.child0.child0, col0.child1, col1, col2, col2.child0, col2.child1]`

We need to advance the iterator by the total number of descendants for a given column. There's no way to know this _a priori_ and requires walking the tree of children. We could do this explicitly with a call to [`count_descendents`](https://github.com/rapidsai/cudf/blob/8c657965bf855167d824e4fc16a72d33cbd8203f/cpp/include/cudf/column/column_view.hpp#L511), but that feels wasteful since the recursive calls to `deserialize_columns` are already walking the child tree and I feel we should exploit that.

File: cpp/src/copying/pack.cpp [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  }
+}
+
+} // namespace anonymous
+
+packed_table pack(cudf::table_view const& input,
+                  cudaStream_t stream,
+                  rmm::mr::device_memory_resource* mr)
+{
+  contiguous_split_result contiguous_data = std::move(contiguous_split(input, {})[0]);
```
[the remark this answers is not in the document]
Comment (devavret) [reply cpp/src/copying/pack.cpp:42]: Did you mean `.front()`?

[the remark this answers is not in the document]
Comment (jrhemstad) [reply cpp/src/copying/pack.cpp:47]: Yes, somehow my brain always messes up `first/last` with `front/back`.

File: cpp/include/cudf/copying.hpp [hunk 3]
```diff
@@ -444,6 +444,25 @@ std::vector<contiguous_split_result> contiguous_split(cudf::table_view const& in
...
+    };
+
+    std::vector<serialized_column> table_metadata;
+    std::unique_ptr<rmm::device_buffer> table_data;
+};
+
+packed_table pack(cudf::table_view const& input,
+                  rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+contiguous_split_result unpack(packed_table & input,
```
[the remark this answers is not in the document]
Comment (devavret) [reply cpp/include/cudf/copying.hpp:463]: Not true. The table data buffer from `packed_table` is `std::move`d out into the returning `contiguous_split_result`. This is done to reduce memory usage. I can change it to pointer.

[the remark this answers is not in the document]
Comment (jrhemstad) [reply cpp/include/cudf/copying.hpp:463]: This should be a `unique_ptr<packed_table>` then to indicate that the function is taking ownership of the `packed_table`. Furthermore, there will need to be a constructor to allow the user to construct a `packed_table` from the host/device buffer.

[the remark this answers is not in the document]
Comment (devavret) [reply cpp/include/cudf/copying.hpp:463]: A `packed_table` should only be allowed to be generated by `cudf::pack`. Unless you mean there should be a method to convert `packed_table` returned by `cudf::pack` into a `unique_ptr<packed_table>`. Then we're just transferring the take-by-ref API design from `unpack` to that method.

There's two more ways this can be handled:
1. Return `unique_ptr<packed_table>` from `pack`.
2. Take by `packed_table ` by rvalue reference in `unpack`.

[the remark this answers is not in the document]
Comment (jrhemstad) [reply cpp/include/cudf/copying.hpp:463]: > A packed_table should only be allowed to be generated by cudf::pack

I think you're forgetting about when sending a `packed_table` over the wire. On the receive side, all there is is a host buffer/device_buffer. The receiver needs to be able to construct a `packed_table` from those buffers, and then pass ownership of those buffers to `cudf::unpack`.

File: cpp/src/copying/pack.cpp (hunk 1 above)
Comment (devavret) [reply cpp/src/copying/pack.cpp:96]: Even with `count_descendants`, we need the tree already generated. Which isn't generated until we actually deserialize.

File: cpp/src/copying/pack.cpp [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+                                                 rmm::device_buffer const& table_data)
+{
+  auto all_data_buffer_ptr = static_cast<uint8_t const*>(table_data.data());
+  size_t data_offset = col.data<uint8_t>() - all_data_buffer_ptr;
+  // TODO: fix for non-nullable
+  size_t null_mask_offset = reinterpret_cast<uint8_t const*>(col.null_mask()) - all_data_buffer_ptr;
+  return packed_table::serialized_column{col.type(), col.size(), data_offset, null_mask_offset, col.num_children()};
+}
+
+void add_columns(std::vector<column_view> const& cols,
```
[the remark this answers is not in the document]
Comment (devavret) [reply cpp/src/copying/pack.cpp]: I had to change the API to return a vector of bytes. For that, and to reduce copying, I now have to pass `std::vector<uint8_t>` around which doesn't translate well into this iterator based scheme as the type of output is not `serialized_column` anymore. In light of that, can you reconsider this comment?

File: cpp/src/copying/pack.cpp (hunk 1 above)
Comment (devavret) [reply cpp/src/copying/pack.cpp]: I refactored this to use a recursive lambda which captures the counter and avoids passing it as in/out parameter.

File: cpp/src/copying/pack.cpp [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+}  // namespace
+
+contiguous_split_result unpack(std::unique_ptr<packed_table> input) {
+  auto serialized_columns =
+    reinterpret_cast<serialized_column const*>(input->table_metadata->data());
+  uint8_t const* base_ptr = static_cast<uint8_t const*>(input->table_data->data());
+  size_t current_index    = 1;
+
+  std::function<std::vector<column_view>(size_type)> get_columns;
```
Comment (jrhemstad) [inline cpp/src/copying/pack.cpp:110]: Why a `std::function` instead of a lambda?

Comment (devavret) [reply cpp/src/copying/pack.cpp:107]: Because `auto` type deduction doesn't work when the function is called recursively. https://stackoverflow.com/questions/2067988/recursive-lambda-functions-in-c11

File: cpp/src/copying/contiguous_split.cu [hunk 6]
```diff
@@ -430,7 +416,7 @@ contiguous_split_result alloc_and_copy(cudf::table_view const& t,
```
Comment (codereport) [inline cpp/src/copying/contiguous_split.cu:416]: After writing the message below, I realized that this isn't a `for_each` - we are calculating `total_size` - this is an `accumulate` - or even better seeing as this is a `.cu` file, `thrust::transform_reduce` with a `thrust::counting_iterator`. Code should be something like:
```cpp
  auto total_size = std::transform_reduce(
    t.begin(), 
    t.end(), 
    thrust::make_counting_iterator(0),
    size_t{0},
    thrust::plus{},
    [&split_info](cudf::column_view const& c, size_type column_index) {
      return cudf::experimental::type_dispatcher(
        c.type(), column_buffer_size_functor{}, c, split_info[column_index]);
    });
```

==== 
I realize you didn't modify the code above but I am interested in people's thought on this. If I need a modifiable value in a lambda passed to an algorithm, I typically mark the lamba as `mutable` and just init the variable by value instead of passing by reference so instead of this:
```cpp
  size_t total_size      = 0;
  size_type column_index = 0;
  std::for_each(
    t.begin(), t.end(), [&total_size, &column_index, &split_info](cudf::column_view const& c) {
      total_size += cudf::experimental::type_dispatcher(
        c.type(), column_buffer_size_functor{}, c, split_info[column_index]);
      column_index++;
    });
```
you have something like this:
```cpp
  std::for_each(
    t.begin(), t.end(), [total_size = 0, column_index = 0, &split_info](cudf::column_view const& c) mutable {
      total_size += cudf::experimental::type_dispatcher(
        c.type(), column_buffer_size_functor{}, c, split_info[column_index]);
      column_index++;
    });
```

**Comment by harrism** [comment]
@devavret is this still active?

**Comment by devavret** [comment]
> @devavret is this still active?

Strictly back burner for now. @shwina and @VibhuJawa tested it and found that in some cases, the performance actually decreased due to the increased memory requirement.

**Comment by harrism** [comment]
@devavret is this PR needed any more?

**Comment by kkraus14** [comment]
Superseded by #7096
